File: app/schemas/response.py
"""
API响应模型

定义统一的API响应格式
"""

from typing import Any, Generic, Optional, TypeVar

from pydantic import BaseModel, Field

T = TypeVar("T")


class BaseResponse(BaseModel, Generic[T]):
    """
    基础响应模型

    所有API响应都应使用此格式
    """

    success: bool = Field(..., description="请求是否成功")
    message: str = Field(..., description="响应消息")
    data: Optional[T] = Field(None, description="响应数据")
    detail: Optional[Any] = Field(None, description="详细信息")


class SuccessResponse(BaseResponse[T]):
    """
    成功响应模型

    用于返回成功响应的API端点
    """

    success: bool = Field(default=True, description="请求是否成功")


class ErrorResponse(BaseResponse[None]):
    """
    错误响应模型

    用于返回错误响应的API端点
    """

    success: bool = Field(default=False, description="请求是否成功")
    status_code: int = Field(..., description="HTTP状态码")


class PaginationMeta(BaseModel):
    """
    分页元数据
    """

    page: int = Field(..., description="当前页码", ge=1)
    page_size: int = Field(..., description="每页数量", ge=1)
    total: int = Field(..., description="总记录数", ge=0)
    total_pages: int = Field(..., description="总页数", ge=0)

    @classmethod
    def create(cls, page: int, page_size: int, total: int) -> "PaginationMeta":
        """
        创建分页元数据

        Args:
            page: 当前页码
            page_size: 每页数量
            total: 总记录数

        Returns:
            PaginationMeta: 分页元数据实例
        """
        total_pages = (total + page_size - 1) // page_size if total > 0 else 0
        return cls(
            page=page,
            page_size=page_size,
            total=total,
            total_pages=total_pages,
        )


class PaginatedResponse(BaseResponse[list[T]]):
    """
    分页响应模型

    用于返回分页数据的API端点
    """

    success: bool = Field(default=True, description="请求是否成功")
    data: list[T] = Field(default_factory=list, description="数据列表")
    meta: Optional[PaginationMeta] = Field(None, description="分页元数据")

File: src/schemas/response.ts
import { z } from "zod";

/**
 * 基础响应模型
 */
export const BaseResponseSchema = <T extends z.ZodTypeAny>(dataSchema: T) =>
  z.object({
    success: z.boolean(),
    message: z.string(),
    data: dataSchema.optional(),
    detail: z.unknown().optional(),
  });

/**
 * 成功响应模型
 */
export const SuccessResponseSchema = <T extends z.ZodTypeAny>(dataSchema: T) =>
  BaseResponseSchema(dataSchema).extend({
    success: z.literal(true),
  });

/**
 * 错误响应模型
 */
export const ErrorResponseSchema = z.object({
  success: z.literal(false),
  message: z.string(),
  detail: z.unknown().optional(),
  status_code: z.number(),
});

/**
 * 分页元数据模型
 */
export const PaginationMetaSchema = z.object({
  page: z.number().int().min(1),
  page_size: z.number().int().min(1),
  total: z.number().int().min(0),
  total_pages: z.number().int().min(0),
});

/**
 * 创建分页元数据
 */
export const createPaginationMeta = (
  page: number,
  pageSize: number,
  total: number
) => {
  const totalPages = total > 0 ? Math.ceil(total / pageSize) : 0;
  return {
    page,
    page_size: pageSize,
    total,
    total_pages: totalPages,
  };
};

/**
 * 分页响应模型
 */
export const PaginatedResponseSchema = <T extends z.ZodTypeAny>(
  itemSchema: T
) =>
  SuccessResponseSchema(z.array(itemSchema)).extend({
    meta: PaginationMetaSchema.optional(),
  });

/**
 * 响应类型定义
 */
export type SuccessResponse<T> = {
  success: true;
  message: string;
  data?: T;
  detail?: unknown;
};

export type ErrorResponse = {
  success: false;
  message: string;
  detail?: unknown;
  status_code: number;
};

export type PaginationMeta = z.infer<typeof PaginationMetaSchema>;

export type PaginatedResponse<T> = {
  success: true;
  message: string;
  data: T[];
  meta?: PaginationMeta;
};

/**
 * 创建成功响应
 */
export const createSuccessResponse = <T>(
  message: string,
  data?: T
): SuccessResponse<T> => {
  return {
    success: true,
    message,
    data,
  };
};

/**
 * 创建分页成功响应
 */
export const createPaginatedResponse = <T>(
  message: string,
  data: T[],
  meta: PaginationMeta
): PaginatedResponse<T> => {
  return {
    success: true,
    message,
    data,
    meta,
  };
};

/**
 * 创建错误响应
 */
export const createErrorResponse = (
  message: string,
  statusCode: number,
  detail?: unknown
): ErrorResponse => {
  return {
    success: false,
    message,
    detail,
    status_code: statusCode,
  };
};
